### PR TITLE
WIP - Implicitly discover schema and query without macro attributes

### PR DESCRIPTION
**Pull request by fnune**
So, this is the most naïve implementation I could fathom. Current problems:

1. `to_snake_case` won't take care of numbers. In the existing example, `Query1` is snake-cased as `query1` but the file name was `query_1`. I suppose this is not a big issue.
2. `implicitly_discover_schema_name` is extremely verbose, I suppose you can come up with some nicer way of writing this.
3. This assumes there's `src/`. There should probably be another approach, and I'd like to store those magic strings somewhere nicer.
4. I'm probably using some archaic way of manipulating strings.

### CODE REVIEW

**Comment by brainlessdeveloper** [comment]
https://github.com/tomhoule/graphql-client/issues/57

File: graphql_query_derive/src/lib.rs [hunk 1]
```diff
@@ -66,6 +66,26 @@ fn read_file(
...
+fn implicitly_discover_schema_name(cargo_manifest_dir: &str) -> Result<String, failure::Error> {
+    if ::std::path::Path::new(&cargo_manifest_dir).join("src/schema.json").exists() {
+        return Ok("src/schema.json".to_owned());
+    }
+    if ::std::path::Path::new(&cargo_manifest_dir).join("src/schema.graphql").exists() {
+        return Ok("src/schema.graphql".to_owned());
+    }
+    if ::std::path::Path::new(&cargo_manifest_dir).join("src/schema.gql").exists() {
+        return Ok("src/schema.gql".to_owned());
+    }
```
Comment (brainlessdeveloper) [inline graphql_query_derive/src/lib.rs]: I'm sure there's a much more elegant way of writing this, but I'm a newb so I'll gladly take suggestions.

**Comment by scrogson** [comment]
2. How about `infer_schema_name`?

File: graphql_query_derive/src/lib.rs (hunk 1 above)
Comment (scrogson) [reply graphql_query_derive/src/lib.rs]: What about something like this?

```rust
fn find_schema_file(cargo_manifest_dir: &str) -> Result<String, failure::Error> {
    let default_paths = vec!["src/schema.json", "src/schema.graphql", "src/schema.gql"];

    default_paths
        .iter()
        .map(|path| Path::new(&cargo_manifest_dir).join(path))
        .find(|path| path.exists())
        .and_then(|path| path.to_owned())
        .ok_or_else(|| Err(format_err!("None of the following exist in src/: schema.json, schema.graphql, schema.gql")))
}
```

Comment (brainlessdeveloper) [reply graphql_query_derive/src/lib.rs]: That looks much nicer! Thanks

**Comment by brainlessdeveloper** [comment]
I just realized I should be parsing `.graphql` for the schema path.

**Comment by scrogson** [comment]
> I just realized I should be parsing .graphql for the schema path.

Do you mean `. graphqlconfig`?

File: graphql_query_derive/src/lib.rs (hunk 1 above)
Comment (tomhoule) [reply graphql_query_derive/src/lib.rs]: +1 for this solution. I'm not sure but I think it can even be an array literal instead of `vec!`, so `let paths = &["src/schema.json", "src/schema.graphql", "src/schema.gql"];`, which would be a tiny bit more efficient but we really don't care in that context since it only runs once per query at compile-time.

**Comment by tomhoule** [comment]
Really nice, thanks for the PR! There are two things I'm not sure about myself, so it's open for discussion:

1. Is `src/schema.{json,graphql}` a reasonable default? In the issue I mentioned maybe using a schema from `.graphqlconfig` since it is a standard for the JS GraphQL tooling. Having the schema at the project root, next to the Cargo.toml could also be a reasonable default.
2. Should we make a separate test for that (with a tiny query and schema since we only want to check that the files are found) in the `tests/` dir or just use the github examlpe as a test?

For queries I think it's fine this way, we shouldn't worry about case conversion edge-cases since there's always the explicit `query_path` option as a fallback.

**Comment by tomhoule** [comment]
see https://github.com/prismagraphql/graphql-config for `.graphqlconfig`

File: graphql_query_derive/src/lib.rs [hunk 2]
```diff
@@ -75,8 +95,14 @@ fn impl_gql_query(input: &syn::DeriveInput) -> Result<TokenStream, failure::Erro
     let cargo_manifest_dir =
         ::std::env::var("CARGO_MANIFEST_DIR").expect("CARGO_MANIFEST_DIR env variable is defined");
 
-    let query_path = extract_attr(input, "query_path")?;
-    let schema_path = extract_attr(input, "schema_path")?;
+    let module_name = Ident::new(&input.ident.to_string().to_snake_case(), Span::call_site());
+    let struct_name = &input.ident;
+
+    let query_path = extract_attr(input, "query_path")
+        .or_else(|_err| implicitly_discover_query_name(&struct_name.to_string()))?;
```
Comment (tomhoule) [inline graphql_query_derive/src/lib.rs]: There's an error message line 60 in this file that gets printed if the file isn't found, we should update it with info on how we try to find the file explicitly, otherwise it will be confusing. The problem is that in `read_file` we don't know whether the path was specified implicitly or just inferred so we can't adapt the error message. Do you have an idea? I'll try to think of something too.

File: graphql_query_derive/src/lib.rs (hunk 1 above)
Comment (scrogson) [reply graphql_query_derive/src/lib.rs]: Yeah, I believe that should work as well. 👍

**Comment by scrogson** [comment]
> Having the schema at the project root, next to the Cargo.toml could also be a reasonable default.

I think I agree with this.

**Comment by brainlessdeveloper** [comment]
So for the schema we could follow something like this:

0. Use the macro attribute.
1. Try to parse `.graphqlconfig` and find it from there (maybe it's worth trying to have a file dedicated to handling stuff from `.graphqlconfig`?).
2. If it's not there, try to find it in the project root next to `Cargo.toml`.
3. Otherwise try to find it in `${root}/src/`.
4. Fail. 

For the query files, I'm thinking defaulting to `${root}/src/` is not the best idea, because most projects will have query files somewhere else in the code. Maybe (if not too magic) something like this would work:

0. Use the macro attribute.
1. Try to find the file searching the directory from which the macro was called, and recurse parent directories until the project root. Maybe also take into account `includes` and `excludes` from `.graphqlconfig`? Not sure if this would be worth it.
2. Fail.

In any case, we could just start by using the defaults you've suggested. What do you think?

**Comment by tomhoule** [comment]
For the schema I agree, as long as we properly document the implicit behaviour so it doesn't feel too magic.

For the queries I was thinking we should look for the query file next to the file where the struct is located (same directory)  but it turns out it's not easy. I am wondering about another feature we want (multiple operations per query document: https://github.com/tomhoule/graphql-client/issues/31). We could use the struct name to easily select an operation inside a query file (as a fallback from an explicit setting or finding the single query in the file) instead, and now it seems more natural to me. What do you think?

**Comment by tomhoule** [comment]
For `.graphqlconfig` yes it's better not to do that in `lib.rs`. But maybe it can be done in another PR if that's too much for this one (I don't know how much work that could be).

**Comment by brainlessdeveloper** [comment]
Updates:

https://crates.io/crates/graphql-config
https://github.com/brainlessdeveloper/graphql-config

**Comment by brainlessdeveloper** [comment]
Does it still make sense to pursue this after so much development has happened on master?

**Comment by tomhoule** [comment]
It would be hard to rebase I think. And nobody has requested the feature so far, so I don't know if it should be pursued. Sorry :/

**Comment by tomhoule** [comment]
@fnune the library changed _a lot_ since this PR was open, I think we should start from scratch if we still want to do something like this. I'm working on graphql-client a bit again, so we can get ideas going in issues. I have a few for a better codegen workflow.

**Comment by fnune** [comment]
Cool! What are the plans?

**Comment by tomhoule** [comment]
I want to completely decouple `graphql_client_codegen` from the compiler's macro architecture, since code generation is based on the GraphQL query + schema, not rust code. That would make it easy to have an in-browser playground that shows you exactly what gets generated.

For the "auto-discovery", the idea would rather be a "batch generation" macro that generates all the modules for your query at once, so you don't need to write the paths of query files as often.
